docs/proposals/network-policy.md

```diff
@@ -0,0 +1,127 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG.  In general,
+egress is expected to augment this proposal rather than modify it, with one exception (noted below).
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta NetworkPolicy API object
+- A field on the `Namespace` object to control isolation.
+
+### Namespace changes
+
+> Note that this section is subject to some change with the addition of egress policy as mentioned above.
+
+The `Namespace` object will be augmented with the field `networkIsolation` which will take one of the following values:
+
+When `networkIsolation=yes` is set on a namespace:
+- Pods in that Namespace will not be accessible from any other source, unless explicitly allowed by a NetworkPolicy object.
+- Pods in that Namespace will be able to access any other source (egress is uninhibited).
+
+When `networkIsolation=no` is set on a namespace:
+- Pods in that Namespace will be accessible from any other source.
+- Pods in that Namespace will be able to access any other source.
+
+The `networkIsolation` field will default to `false` if not specified.
+
+### NetworkPolicy YAML Definition and Behavior
+
+Once the namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+a NetworkPolicy should allow any incoming connections that match any
+ingress rule, per the network plugin’s capabilities.
+
+```yaml
+kind: NetworkPolicy
+apiVersion: v1beta 
+metadata:
+  name:
+  namespace:
+spec:
+  podSelector:            // Standard label selector - selects pods.  
+  ingress:                // List of ingress rules (optional).             
```

Comment:
It's the cross product of label combinations that blows it up, a simple example:
You have pods `app=A` => `app=B` => `app=C` and you use the labels `env={qa|prod|dev}` and `tenant={Birds|Lizards}` to isolate the pods in a multi tenant environment. Maybe not the best example but it really applies to any scenario where you leverage multiple labels to isolate things from each other.

This would result in the following individual policy elements to be required to enforce it:
```
selector: app=B, env=qa, tenant=coke
allow: app=A, env=qa, tenant=coke
selector: app=C, env=qa, tenant=coke
allow: app=B, env=qa, tenant=coke
selector: app=B, env=prod, tenant=coke
allow: app=A, env=prod, tenant=coke
selector: app=C, env=prod, tenant=coke
allow: app=B, env=prod, tenant=coke
selector: app=B, env=dev, tenant=coke
allow: app=A, env=dev, tenant=coke
selector: app=C, env=dev, tenant=coke
allow: app=B, env=dev, tenant=coke
selector: app=B, env=qa, tenant=pepsi
allow: app=A, env=qa, tenant=pepsi
selector: app=C, env=qa, tenant=pepsi
allow: app=B, env=qa, tenant=pepsi
selector: app=B, env=prod, tenant=pepsi
allow: app=A, env=prod, tenant=pepsi
selector: app=C, env=prod, tenant=pepsi
allow: app=B, env=prod, tenant=pepsi
selector: app=B, env=dev, tenant=pepsi
allow: app=A, env=dev, tenant=pepsi
selector: app=C, env=dev, tenant=pepsi
allow: app=B, env=dev, tenant=pepsi
```